### DISCUSSION TITLE
OnPageChange behavior for same URL path RFC

### DISCUSSION THREAD

**Discussion by dillonkearns** [dillonkearns/elm-pages#184] [Ideas]
The OnPageChange Msg allows users to listen for page change events since `elm-pages` wraps the `Browser.application` and handles routing.

What should happen in these two cases?

1) You navigate to a URL with the same path as your current location, but it has a different fragment (example: start at URL `/docs/getting-started` and click a link to `/docs/getting-started#cli`)
2) You navigate to the exact same URL (for example, you're at `/docs` and you click a link to `/docs`)

It seems like the user should have *some way* to know about this event. So the obvious solution is to just fire `OnPageChange` in both cases. Because otherwise the user would simply have no way to handle those events.

## `init` and `Cmd`s for these cases

What about `init`? In cases (1) and (2), should `init` be triggered, causing the page's Model to be started from scratch, and firing off any user Cmd's for the page? This would mean that clicking to a URL fragment would cause a new `init`, for example if you click a link in the table of contents to take you to a new section. That feels a little strange in that case.

One possible direction to explore is to give the user a way to call `init` in `Shared.elm`.

```elm
onPageChange : Url -> Url -> model -> ( model, Cmd msg ) -> ( model, Cmd msg )
onPageChange previousUrl newUrl currentModel ( initModel, initCmd ) =
    ( initModel, initCmd )
```

Then you could define custom logic for how to handle those situations in your app. You could even trigger an event on page change.

That's pretty low-level, though, and it would be ideal to just "do the right thing" if there's a way to get intuitive behavior without exposing low-level details like that.

**Comment by MarcoDaniels** [comment]
I like the idea of having this event on every page navigation even when it is a hash link `/url#something` as it is gives the ability to trigger some custom event (can be animation, tracking, etc.).

I just feel like the name `OnPageChange` might be misleading if it gets triggered when the user is on page `/docs` and clicks `/docs` as technically the page didn't really change but it might have reloaded/refreshed. 

Not sure exactly what name I would give though, but maybe something like `onNavigation`, `onPageLoad`? I also don't believe it to be a dealbreaker.
